src/github.com/getlantern/flashlight/ui/service.go

```diff
@@ -0,0 +1,102 @@
+package ui
+
+import (
+	"encoding/json"
+	"sync"
+)
+
+type envelopeType struct {
+	Type string
+}
+
+type Service struct {
+	Name    string
+	In      chan []byte
+	Out     chan []byte
+	helloFn func(func([]byte) error) error
+}
+
+var (
+	mu               sync.Mutex
+	defaultUIChannel *UIChannel
+
+	out      = make(chan []byte, 10)
+	services = make(map[string]*Service)
+)
+
+func (s *Service) watch() {
+	// Watch for new messages and sent them to the combined output.
+	for b := range s.Out {
+		out <- b
+	}
+}
+
+func Register(name string, helloFn func(func([]byte) error) error) (*Service, error) {
+	mu.Lock()
+
+	if defaultUIChannel == nil {
+		// Don't start until a service is registered.
+		start()
+	}
+
+	if services[name] != nil {
+		// Using panic because this would be a developer error rather that
+		// something that could happen naturally.
+		panic("Service was already registered.")
+	}
+
+	services[name] = &Service{
+		Name: name,
+		In:   make(chan []byte, 10),
+		// We should probably use a buffered channel.
+		Out:     make(chan []byte),
+		helloFn: helloFn,
+	}
+
+	go services[name].watch()
+
+	mu.Unlock()
+
+	return services[name], nil
+}
+
+func start() {
+	// Establish a channel to the UI for sending and receiving updates
+	defaultUIChannel = NewChannel("/data", func(write func([]byte) error) error {
+		// Sending hello messages.
+		for _, s := range services {
+			// Delegating task...
+			if err := s.helloFn(write); err != nil {
+				log.Errorf("Error writing to socket: %q", err)
+			}
+		}
+		return nil
```

Comment:
What happens if a service is registered after a client has already connected?  Seems like you should process the helloFn at that point, right?